### PR TITLE
test(waku-filter): Unsubscribe tests

### PR DESCRIPTION
**Pull request by AlejandroCabeza**
# Description
Implement unsubscribe tests for waku filter.

# Changes
- [X] Implement waku filter unsubscribe tests.
- [X] Update waku filter protocol so it returns an error on wrong topic parameters.

# Dependencies
https://github.com/waku-org/nwaku/pull/2057

### CODE REVIEW

**Review by NagyZoltanPeter — APPROVED** [review]
Awesome extensive testing! Thank you!
Only one suggestion for subscription exclusion use.

File: waku/waku_filter_v2/protocol.nim [hunk 1]
```diff
@@ -86,7 +86,11 @@ proc unsubscribe(wf: WakuFilter, peerId: PeerID, pubsubTopic: Option[PubsubTopic
     return err(FilterSubscribeError.notFound())
 
   var peerSubscription = wf.subscriptions.mgetOrPut(peerId, initHashSet[FilterCriterion]())
-  # TODO: consider error response if filter criteria does not exist
+
+  if not peerSubscription.containsAny(filterCriteria):
+    debug "unsubscribing peer is not subscribed to any of the content topics in this pubsub topic", peerId=peerId, pubsubTopic=pubsubTopic.get(), contentTopics=contentTopics
+    return err(FilterSubscribeError.notFound())
+
   peerSubscription.excl(filterCriteria)
```
[the remark this answers is not in the document]
Comment (AlejandroCabeza) [reply waku/waku_filter_v2/protocol.nim:94]: That method is not compliant with peerSubscription/filterCriteria
```
waku/waku_filter_v2/protocol.nim(91, 20) Error: type mismatch: got <HashSet[subscriptions.FilterCriterion], FilterCriteria>
but expected one of:
proc missingOrExcl[A](s: var HashSet[A]; key: A): bool
  first type mismatch at position: 1
  required type for s: var HashSet[missingOrExcl.A]
  but expression 'filterCriteria' is immutable, not 'var'
proc missingOrExcl[A](s: var OrderedSet[A]; key: A): bool
  first type mismatch at position: 1
  required type for s: var OrderedSet[missingOrExcl.A]
  but expression 'filterCriteria' is of type: HashSet[subscriptions.FilterCriterion]

expression: missingOrExcl(filterCriteria, peerSubscription)
```

The second part, about requiring an OrderedSet I think I could solve it, the first part I don't.

[the remark this answers is not in the document]
Comment (NagyZoltanPeter) [reply waku/waku_filter_v2/protocol.nim:94]: Ah, you're right. Anyhow thanks for the effort.

File: waku/waku_filter_v2/subscriptions.nim [hunk 2]
```diff
@@ -43,3 +43,11 @@ proc removePeers*(subscriptions: var FilterSubscriptions, peerIds: seq[PeerID])
...
   for peerId in peerIds:
     subscriptions.removePeer(peerId)
+
+proc containsAny*(criteria: FilterCriteria, otherCriteria: FilterCriteria): bool =
+  ## Check if a given pubsub topic is contained in a set of filter criteria
+  ## TODO: Better way to achieve this?
+  for criterion in otherCriteria:
+    if criterion in criteria:
+      return true
+  false
```
Comment (Ivansete-status) [inline waku/waku_filter_v2/subscriptions.nim:53]: Maybe we could make something like the following:
```suggestion
  let interesctionResult = criteria * otherCriteria
  return interesctionResult.len > 0
```

I don't know which option is faster. If possible, we could benchmark that in a separate tiny Nim example.

**Review by Ivansete-status — APPROVED** [review]
LGTM! Thanks!
Just added a little suggestion

File: waku/waku_filter_v2/subscriptions.nim (hunk 2 above)
Comment (AlejandroCabeza) [reply waku/waku_filter_v2/subscriptions.nim:53]: Just ran a test, and I get the for-if solution is 5~ times faster than the intersection one. I guess we're going with the first 😆

**Pull request merged**